data/blog/2025/2025-08-06-strangler-fig-pattern.md

```diff
@@ -0,0 +1,139 @@
+
+## How Strangler Fig Pattern Works: Step-by-Step
+
+### Step 1: Understand and Isolate Functionality
+
+Analyze the legacy system to identify well-defined boundaries or modules.
+Select a coherent part of the functionality that can be isolated and replaced independently (e.g. user authentication, payment processing).
+It's a good idea to work with the original developers for this step, if they are still available.
+
+### Step 2: Build the Replacement Code
+
+Implement the isolated functionality in the new platform using modern architectures (e.g. microservices, middleware).
+Ensure that the new service meets all the functional and non-functional requirements (performance, security etc.).
+
+### Step 3: Intercept Calls Via the Routing Layer
+
+Use a routing mechanism (like an API gateway, proxy, service mesh) to redirect calls that use that functionality to the new service.
+This allows the rest of the application to continue working as-is while the new component handles specific operations.
+
+### Step 4: Test and Monitor
+
+Thoroughly test the new service in production (with e.g. canary releases, A/B testing).
+Monitor logs, performance metrics and error rates to validate correctness and stability.
+
+### Step 5: Repeat
+
+Repeat the process from steps 1-4 for each component of the legacy system that needs to be replaced.
+Each new component replaces a legacy component, meaning the latter is not used and can be removed.
+
+### Step 6: Retire the Legacy System
+
+Once all functionality is handled by the new services, the legacy system becomes obsolete.
+You can fully decommission and remove the old system.
+
+## Example in Practice
+
+Consider this scenario:
+
+You have a monolithic e-commerce application built with Laminas MVC.
+It has several functional modules:
+
+- `/products` - Product listings `/product/list`, editing `/product/edit` and deleting  `/product/delete`
+- `/cart` - Cart management
+- `/checkout` - Payments and orders
+- `/account` - User account management
+
+You want to migrate to a middleware-based architecture using Mezzio microframewok with Laminas components.
+
+Let's follow the Strangler Fig Pattern in action:
+
+- **Install the Mezzio microframework skeleton**: This is where the new code will be written as microservices.
+- **Select a feature**: Start with `/product/list`, a non-critical module that is easy to isolate.
+- **Build a new microservice**: Implement microservices for product listing, editing and deleting in the middleware platform.
+- **Intercept requests**: Use an API gateway to route `/product/list` calls to the new microservice.
+- **Test and monitor**: Gradually shift traffic from the monolith to the microservice.
+- **Continue migrating**: Repeat the process for each feature like `/product/list`, `/cart` etc., one at a time.
+- **Turn off the monolith**: Once all features are replaced, shut down the legacy application.
```

Comment:
Don't introduce a new term ("monolith") here. Just say "Turn off the legacy application" or "Retire the legacy application"